shadowsocks/cipher_list.go

```diff
@@ -16,12 +16,71 @@ package shadowsocks
 
 import (
 	"container/list"
+	"fmt"
 	"net"
 	"sync"
 
 	"github.com/shadowsocks/go-shadowsocks2/shadowaead"
 )
 
+// All ciphers must have a nonce size this big or smaller.
+const maxNonceSize = 12
+
+// Describes the size parameters of a cipher.
+type cipherSize struct {
+	salt, overhead int
+}
+
+// These are cipher size parameters supported by CipherList.
+// To support TCP trial decryption, these sizes must not include
+// too wide a range.  Otherwise, trial decryption for a large-size
+// cipher could block indefinitely, when the client has actually
+// sent a handshake for a short cipher.
+//
+// Ciphers do not need to have exactly the same size because a minimal
+// handshake (from a short cipher) includes at least one additional
+// AEAD tag, providing enough bytes for trial decryption with a
+// longer cipher.
+var supportedSizes = [...]cipherSize{
+	{32, 16}, // chacha20poly1305, AES-256-GCM
+	{24, 16}, // AES-192-GCM
+	{16, 16}, // AES-128-GCM
+}
+
+// Minimum number of bytes to read for trial decryption of all supported
+// ciphers.
+var tcpHeader int
```

Comment:
Can we calculate this for the specific CipherList? Maybe have it as a member of the class.

Reply:
Done